File: Linked-Lists/Challenge-2.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const processLinkedList = (head) => {
  let current = head;
  while (current !== null) {
    console.log(current.val);
    current = current.next;
  }
};

const createLinkedList = (values) => {
  if (values.length === 0) {
    return null;
  }
  let head = new Node(values[0]);
  let current = head;

  for (let i = 1; i < values.length; i++) {
    let newNode = new Node(values[i]);
    current.next = newNode;
    current = newNode;
  }
  return head;
};

const values = [1, 2, 3, 4, 5, 6, 7];
const head = createLinkedList(values);

// Time Complexity: O(N) And Space Complexity: O(1)

const returnKElement = (head, k) => {
  let current = head;
  let lastNodeCount = 0;
  let currentNodeCount = 0;

  while (current !== null) {
    lastNodeCount += 1;
    current = current.next;
  }
  current = head;
  while (current !== null) {
    currentNodeCount += 1;
    if (lastNodeCount - currentNodeCount === k - 1) {
      console.log(current.val);
      return current.val;
    }
    current = current.next;
  }
};

returnKElement(head, 2);

// Time Complexity: O(N) And Space Complexity: O(1)

const returnKElement2 = (head, k) => {
  let pointerOne = head;
  let pointerTwo = head;

  for (let i = 0; i < k; i++) {
    pointerOne = pointerOne.next;
  }
  while (pointerOne !== null) {
    pointerOne = pointerOne.next;
    pointerTwo = pointerTwo.next;
  }
  console.log(pointerTwo.val);
  return pointerTwo.val;
};

returnKElement2(head, 2);

File: Linked-Lists/Challenge-2.py
class Node:
  def __init__(self, val) -> None:
    self.val = val
    self.next = None


def print_linked_list(head):
  current = head
  while current is not None:
    print(current.val)
    current = current.next

def create_linked_list(values):
  head = Node(values[0])
  current = head

  for value in values[1:]:
    new_node = Node(value)
    current.next = new_node
    current = new_node

  return head

values = ['1','2','3','4','5','6']
head = create_linked_list(values)

# 0(N) Time Complexity and 0(1) Space Complexity

def return_k_element(head,k):
  pointer_one = head
  pointer_two = head

  for i in range(k):
    pointer_one = pointer_one.next

  while pointer_one is not None:
    pointer_one = pointer_one.next
    pointer_two = pointer_two.next

  return pointer_two.val


result = return_k_element(head, 2)
print(result)
